adding sourcemaps so that ml5.js is more readable.

oveddan opened the pull request:
switching to modern browser presets env.

This essentially makes it so that if you copy ml5.js and ml5.js.map you will have the full original source files, making it easier to debug.

It would be worth considering a different `source-map` option that is faster.

oveddan commented:
Do not merge this yet...I'm working on an implementation in rollup.

cvalenzuela commented:
I have used rollup a few times but I'm not super familiar with it. Do you think it is worth the effort to migrate to rollup instead of babel/webpack? I've never used if for buiding a library.  It seems easier to set up and will allows for tree shaking (althought not sure if this something useful for begginers), no?

oveddan commented:
I'm not that familiar with rollup either...but just looking around github, people have seemed to have switched to it (see redux and react for example).  tfjs uses it too.

One thing it does well is remove parts of the code from libraries that are not imported, which in the end makes the bundle size a lot smaller.  In our case when we are including so many deps it's worth having this in there.

Anyways if this doesn't work in the end, sticking with webpack should be fine.

cvalenzuela commented:
I'm down to migrate if necessary. Let's wait for https://github.com/tensorflow/tfjs/issues/393 to be resolved. Although we could move forward with the sourcemaps if you want

shiffman commented:
In terms of priorities, maybe this is something we tackle post beta launch?

oveddan commented:
I agree rollup can wait til after the launch. 

That being said, we should still do something where you can see the source code when developing examples.   Currently it is all minified, which makes it really challenging to debug problems when there is no proper stack trace or the ability to add breakpoints to the ml5 library.  

If we add this...it would actually speed up development of the final features we want before launch as it would make development a lot easier.

shiffman commented:
Agreed! @oveddan if you want to work on something now to fix the issue with `ml5.js` also being minified I agree we should fix that as soon as possible under the current build system. . and revisit the rollup vs webpack question later?

oveddan commented:
Ok will do.  For now will stick to the current build system but just make it not minify ml5.js .

oveddan commented:
I'm also wondering - do we need to be committing ml5.js and ml5.dist.js? I feel like those should be ignored, and will just lead to a lot of git conflicts when people are working in different branches.

.circleci could handle building those and uploading them as packages.

oveddan commented:
Ok so see the current changes to this branch.

I followed the conventions in the [webpack production](https://webpack.js.org/guides/production/) documentation:

* Split webpack config into dev and prod and common, using the webpack-merge to merge them.
*  For dev config, using eval source map.
* For prod config, uglifying and doing a standard source-map that produces a separate file.
* Since we have a ml5.min.js and sourcemap, there is no longer a near for
ml5.js
* Changed the watch task to be yarn start

So now if you reference the script `ml5.min.js`, and `ml5.min.map.js` is in the same directory, you get the full original source code in the dev tools.

Also, a nice discovery through this.  When developing the examples, you can run in this repository:
`yarn start`

And then in the examples, change the script tag to point to:
<script src="http://localhost:8080/ml5.js"></script>

you get live reloading from the original source code in the ml5 library.

cvalenzuela commented:
Amazing, thanks so much @oveddan. we were missing this for a while now!
I'll hopefully merge by today/tomorrow. I'm just going to @handav talk!

cvalenzuela commented on package.json:
```diff
@@ -10,10 +10,8 @@
-    "build": "npm-run-all build:*",
-    "build:main": "webpack --output-filename ml5.js",
-    "build:min": "uglifyjs dist/ml5.js -o dist/ml5.min.js",
+    "start": "webpack-dev-server --open --config webpack.dev.babel.js",
```
The `--open` flag launches a web browser in `localhost:8080` but with nothing on it. It might be better to remove this because it might confuse begginers. @oveddan what do you think is best practice in this case?

oveddan replied:
Yes this stumped me....I think we should remove it unless there is some demo home page.

cvalenzuela replied:
Maybe that is a good idea. Whenever someone does runs `npm start` the browser opens to a page with a few instructions on how to start contributing. If we wanted something like that, where should that file be placed?

oveddan replied:
The instructions are a great idea. I'm not sure yet how to add that but will do the necessary research to figure that out :)

oveddan commented on package.json:
```diff
@@ -75,16 +74,17 @@
-      "es2015",
-      "stage-2"
+      [
+        "env"
```
I don't believe it does. From the [documentation](https://new.babeljs.io/docs/en/next/babel-preset-env.html)

> Without any configuration options, `@babel/preset-env` behaves exactly the same as `@babel/preset-es2015`, `@babel/preset-es2016` and `@babel/preset-es2017` together (or the deprecated babel-preset-latest).

I wouldn't be opposed to adding stage-2, if there are some things we would want to use from there.  As of now, just using `env` seems to have all the features we currently use.

cvalenzuela replied:
let's leave it just with `env` for now and switch if neccessary

oveddan commented:
What do you think about removing "ml5.js and ml5.min.js" from being committed?  I feel like building that can just be taken care of in .circleci

cvalenzuela commented:
Yes, let's not commit those files. We will just need to update the [`README.md`](https://github.com/ml5js/ml5-library/blob/master/README.md#usage) reference to download the library. We where pointing to the `dist`(https://raw.githubusercontent.com/ml5js/ml5-library/master/dist/ml5.js) because it was helpful to have a direct download reference link. Let's replace it with https://unpkg.com/ml5

cvalenzuela commented:
I'm going to merge this now because I'm already using this build changes to push some updates. We are just missing creating an HTML page to help contribute and not committing the `dist` folder. We can open a new PR with those changes!
